app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.kt

```diff
@@ -346,8 +347,15 @@ class RoomActivity : BaseActivity() {
         screenCaptureMenuItem = menu.findItem(R.id.share_screen_menu_item)
         deviceMenuItem = menu.findItem(R.id.device_menu_item)
         requestPermissions()
-        roomViewModel.viewState.observe(this, { roomViewState: RoomViewState -> bindRoomViewState(roomViewState) })
-        roomViewModel.viewEffects.observe(this, { roomViewEffectWrapper: ViewEffect<RoomViewEffect> -> bindRoomViewEffects(roomViewEffectWrapper) })
+
+        onStates(roomViewModel) { state ->
```

Comment:
Just a general question but does `onStates` support generic types? Is there any reason the callback would receive anything other than a RoomViewState?

Reply:
Yes it expects a `UIState` type which we extend in the `RoomViewState` class. I think the idea here is that this allows a `ViewModel` to emit different `UIState` types if you wanted to. I personally think it's easier to stick to a single type for a `ViewModel` which is why I required a single view state type in that `BaseViewModel` class in UDF.